src/asynchronous/client.rs

```diff
@@ -218,3 +159,153 @@ impl Drop for ClientClose {
         trace!("All client is droped");
     }
 }
+
+#[derive(Debug)]
+struct ClientBuilder {
+    rx: Option<MessageReceiver>,
+    streams: Arc<Mutex<HashMap<u32, ResultSender>>>,
+}
+
+impl Builder for ClientBuilder {
+    type Reader = ClientReader;
+    type Writer = ClientWriter;
+
+    fn build(&mut self) -> (Self::Reader, Self::Writer) {
+        let (notifier, waiter) = shutdown::new();
+        (
+            ClientReader {
+                shutdown_waiter: waiter,
+                streams: self.streams.clone(),
+            },
+            ClientWriter {
+                rx: self.rx.take().unwrap(),
+                shutdown_notifier: notifier,
+
+                streams: self.streams.clone(),
+            },
+        )
+    }
+}
+
+struct ClientWriter {
+    rx: MessageReceiver,
+    shutdown_notifier: shutdown::Notifier,
+
+    streams: Arc<Mutex<HashMap<u32, ResultSender>>>,
+}
+
+#[async_trait]
+impl WriterDelegate for ClientWriter {
+    async fn recv(&mut self) -> Option<GenMessage> {
+        self.rx.recv().await
+    }
+
+    async fn disconnected(&self, msg: &GenMessage, e: Error) {
+        // TODO:
+        // At this point, a new request may have been received.
+        let resp_tx = {
+            let mut map = self.streams.lock().unwrap();
+            map.remove(&msg.header.stream_id)
+        };
+
+        // TODO: if None
+        if let Some(resp_tx) = resp_tx {
+            let e = Error::Socket(format!("{:?}", e));
+            resp_tx
+                .send(Err(e))
+                .await
+                .unwrap_or_else(|_e| error!("The request has returned"));
+        }
+    }
+
+    async fn exit(&self) {
+        self.shutdown_notifier.shutdown();
+    }
+}
+
+struct ClientReader {
+    streams: Arc<Mutex<HashMap<u32, ResultSender>>>,
+    shutdown_waiter: shutdown::Waiter,
+}
+
+#[async_trait]
+impl ReaderDelegate for ClientReader {
+    async fn shutdown(&self) {
+        self.shutdown_waiter.wait_shutdown().await
+    }
+
+    async fn disconnected(&self, e: Error, sender: &mut task::JoinHandle<()>) {
+        // Abort the request sender task to prevent incoming RPC requests
+        // from being processed.
+        sender.abort();
+        let _ = sender.await;
+
+        // Take all items out of `req_map`.
+        let mut map = std::mem::take(&mut *self.streams.lock().unwrap());
+        // Terminate outstanding RPC requests with the error.
```

Comment:
What's the mean of `outstanding`? how about undone

Reply:
fixed